File: sorting/merge_sort/merge_sort.cpp
#include <algorithm>
#include <cstddef>
#include <vector>

template<typename T>
void merge_sort(typename std::vector<T>::iterator begin, typename std::vector<T>::iterator end) {
  size_t len = end - begin;

  if (len < 2) {
    return;
  }

  size_t mid = len / 2;

  // 배열을 반으로 나눈 후 각각 재귀적으로 정렬
  merge_sort<T>(begin, begin + mid);
  merge_sort<T>(begin + mid, end);

  // 부분 배열 정렬 결과 임시 배열에 저장
  auto temp = vector(begin, end);

  size_t left = 0;
  size_t right = mid;
  auto iter = begin;

  // 부분 배열 합치기
  while (left < mid && right < len) {
    if (temp[left] <= temp[right]) {
      *iter++ = temp[left++];
    }
    else {
      *iter++ = temp[right++];
    }
  }
  while (left < mid) {
    *iter++ = temp[left++];
  }
  while (right < len) {
    *iter++ = temp[right++];
  }
}

File: sorting/merge_sort/merge_sort.rs
fn merge_sort<T: Ord + Copy>(array: &mut [T]) {
    let len = array.len();

    if len < 2 {
        return;
    }

    let mid = len / 2;

    // 배열을 반으로 나눈 후 각각 재귀적으로 정렬
    merge_sort(&mut array[0..mid]);
    merge_sort(&mut array[mid..len]);
    
    let mut merged = Vec::with_capacity(len);

    let mut left = 0;
    let mut right = mid;

    // 부분 배열 합치기
    while left < mid && right < len {
        if array[left] <= array[right] {
            merged.push(array[left]);
            left += 1;
        }
        else {
            merged.push(array[right]);
            right += 1;
        }
    }
    for i in left..mid {
        merged.push(array[i]);
    }
    for i in right..len {
        merged.push(array[i]);
    }

    // 원래 배열에 합친 결과 넣기 
    array.copy_from_slice(&merged);
}